lib/dns/dns.cc

```diff
@@ -0,0 +1,1114 @@
+	 */
+	static uint32_t gatewayIP = 0;
+
+	/**
+	 * Maximum number of retries for a DNS query. See `perform_dns_lookup`.
+	 */
+	static constexpr const uint8_t DNSMaxRetries = 10;
+
+	/**
+	 * Timeout for one DNS query, in milliseconds. Passed this timeout, we
+	 * will try re-sending a query to the server. RFC 1035 recommends a
+	 * value of 2-5 seconds, i.e., between 2000 and 5000 milliseconds.
+	 */
+	static constexpr const int DNSQueryTimeout = 3000;
+
+	/**
+	 * Send an ARP request to passed local IP.
+	 *
+	 * Note: if our own IP address has not yet been determined, this will
+	 * send an ARP probe.
+	 */
+	void send_arp_request(uint32_t ip)
+	{
+		struct FullARPPacket arpPacket = {0};
+
+		memcpy(&arpPacket.ethernet.source, deviceMAC.data(), 6);
+		memset(&arpPacket.ethernet.destination, 0xff, 6);
+
+		arpPacket.ethernet.etherType = EtherType::ARP;
+		arpPacket.arp.htype          = htons(0x1);
+		arpPacket.arp.ptype          = EtherType::IPv4;
+		arpPacket.arp.hlen           = 0x6 /* size of a MAC address */;
+		arpPacket.arp.plen           = sizeof(uint32_t);
+		arpPacket.arp.oper           = ARPRequest;
+		memcpy(&arpPacket.arp.sha, deviceMAC.data(), 6);
+		// This will be zero if our own IP address has not yet been
+		// determined.
+		arpPacket.arp.spa = deviceIP;
+		arpPacket.arp.tpa = ip;
+
+		ethernet_send_frame(reinterpret_cast<uint8_t *>(&arpPacket),
+		                    sizeof(arpPacket));
+	}
+
+	/**
+	 * Send a DNS query for passed `hostname` of length `length` (not
+	 * including the zero terminator).
+	 */
+	void send_dns_query(const char *hostname, size_t length, bool askIPv6)
+	{
+		Debug::log("Sending a DNS query for {} (IPv6: {})", hostname, askIPv6);
+
+		// DNS query = length of the hostname + 2 (needed for the
+		// encoding of the hostname) + 2 (qtype) + 2 (qclass)
+		size_t packetSize = sizeof(FullDNSPacket) + length + 6;
+
+		uint8_t dnsPacket[packetSize];
+		memset(dnsPacket, 0, packetSize);
+		FullDNSPacket *header = reinterpret_cast<FullDNSPacket *>(dnsPacket);
+
+		// Device (source) MAC.
+		memcpy(&header->ethernet.source, deviceMAC.data(), 6);
+		memcpy(&header->ethernet.destination, dnsServerMAC.data(), 6);
+		// Only support IPv4 for now.
+		header->ethernet.etherType = EtherType::IPv4;
+
+		// 5 x 32 bit = 20 bytes (= IPv4 header length).
+		header->ipv4.versionAndHeaderLength = (4 << 4) | 5;
+		header->ipv4.packetLength = htons(packetSize - sizeof(EthernetHeader));
+		// Default TTL as recommended by RFC 1700.
+		header->ipv4.timeToLive         = 64;
+		header->ipv4.protocol           = IPProtocolNumber::UDP;
+		header->ipv4.sourceAddress      = deviceIP;
+		header->ipv4.destinationAddress = dnsServerIP;
+		// Calculate the checksum last.
+		header->ipv4.headerChecksum = compute_ipv4_checksum(
+		  reinterpret_cast<uint8_t *>(&header->ipv4), sizeof(IPv4Header));
+
+		// Use the DNS server port to originate requests, as we are
+		// sure the TCP/IP stack won't use this one.
```

Comment:
We don't. But I suspect that we want a slightly different check. We don't want the network stack to talk to the DNS server (`$DNS server IP : $DNS server port`). And conversely we don't want the DNS resolver to do anything else than talking to the DNS server or doing ARP.

I think that this would be easy to add (we just have to check the `ForwardFlags` in `packet_filter_egress`), however we would need two versions of `ethernet_send_frame`, one for the TCP/IP stack and one for the DNS resolver.

Would that make sense? Then we don't have to worry about which port to use.

Reply:
Actually I realize that this is harder to implement that I thought, because it is the driver that calls `packet_filter_egress`, at which point we do not know who (TCP/IP stack or DNS resolver) sent the frame.

Reply:
Without this check, the TCP/IP stack can send a packet to the DNS server and impersonate the DNS resolver. But I do not think this has any sort of implication - the DNS resolver would simply receive unsolicited answers and drop them. Things are probably fine as they are now.